sycl/test/usm/badmalloc.cpp

```diff
@@ -0,0 +1,45 @@
+// RUN: %clangxx -fsycl %s -o %t1.out
+// RUN: env SYCL_DEVICE_TYPE=HOST %t1.out
+// RUN: %CPU_RUN_PLACEHOLDER %t1.out
+// RUN: %GPU_RUN_PLACEHOLDER %t1.out
+
+//==----------------- badmalloc.cpp - Bad Mallocs test ---------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+
+// This test verifies that things fail in the proper way when they should.
+
+#include <CL/sycl.hpp>
+
+using namespace cl::sycl;
+
+int main(int argc, char *argv[]) {
+  queue q;
+
+  // Good size, bad type
+  auto p = malloc(8, q, usm::alloc::unknown);
+  assert(p == nullptr);
+
+  // Bad size, host
+  p = malloc(-1, q, usm::alloc::host);
+  assert(p == nullptr);
+
+  // Bad size, device
+  p = malloc(-1, q, usm::alloc::device);
+  assert(p == nullptr);
+
+  // Bad size, shared
+  p = malloc(-1, q, usm::alloc::shared);
+  assert(p == nullptr);
+
+  // Bad size, unknown
+  p = malloc(-1, q, usm::alloc::unknown);
+  assert(p == nullptr);
+
```

Comment:
Should we add test case suggested by Ronan: size = 0, alignment = 1?
BTW, shouldn't we call `aligned_alloc` instead of `malloc`?

Reply:
Added.

Reply:
I mean we probably should replace `malloc` calls with `aligned_alloc`.
This patch changes behavior of `aligned_alloc`, but test calls `malloc`.

Also, note, that checks added for `aligned_alloc` fail.
Probably, `aligned_alloc` implementation still has a bug noted by Ronan. It doesn't return `nullptr` for the case `aligned_alloc(1, 0, ...);`.

Reply:
Malloc = aligned_alloc with 128 byte alignment

Reply:
> Malloc = aligned_alloc with 128 byte alignment

All devices or host only?

Reply:
All devices.

Reply:
I did that at someone's request.  Maybe it was Vlad.

Anyhoo - I rewrote things a bit.

Mallocs of sz 0 are technically ok.

The problem behavior I'm trying to stop is when your aligned_allocator overflows the allocation sz.

The NumBytes massaging turns (size_t)(-1) to 0.  This is bad and should be caught.  Malloc of 0 is technically ok.

Reply:
I still think that testing `aligned_alloc` calling `malloc` is bad idea. If someone change `malloc` implementation and replace `aligned_alloc` with something else, this test will not cover the code it's supposed to cover.
Also it's not possible to test all the inputs via malloc interface. E.g. alignment != 128.

I think it might be a good idea to leave the tests checking that allocating memory of size 0 is valid - i.e. `aligned_alloc`/`malloc` return value is not `nullptr`.

Reply:
Technically malloc of 0 is "implementation defined".

CPU OCL seems to not like it. 
Will add explicit aligned_alloc calls to test.